### ISSUE TITLE
~ is shortened incorrectly when reaching it with cd ..

### ISSUE THREAD

**Issue by iFreilicht** [bhilburn/powerlevel9k#698]
The paths  `/` and `~` are special cases for the directory shortener. They should always appear without a delimiter. This is not the case when using `cd ..` to enter `~`:

![zsh-problem](https://user-images.githubusercontent.com/9742635/33885063-6deb0b9a-df42-11e7-9d02-8b564fec4b0d.png)

All settings regarding the shortening in my `.zshrc`:

```
POWERLEVEL9K_SHORTEN_DIR_LENGTH=1
POWERLEVEL9K_SHORTEN_STRATEGY=Default
POWERLEVEL9K_SHORTEN_DELIMITER="…"
```

My delimiter of choice is the ellipsis character … instead of three dots ..., but this happens with any delimiter, even an empty one.

**Comment by bhilburn** [comment]
Wow, you're right... that is definitely broken.

This should be pretty easy to fix, and it would just be fixing the conditional to properly parse/strip the `cwd` in that segment logic. I'll mark this as a good beginner bug to fix in case anyone is up for it =)

**Comment by iFreilicht** [comment]
I tried my hand at it, and I don't think it's that easy.

The problem is [line 812](https://github.com/bhilburn/powerlevel9k/blob/c50eb9cdda0992d7f717fe5ffaea8812ea5146f9/powerlevel9k.zsh-theme#L812) in powerlevel9k.zsh-theme:

```zsh
current_path="$(print -P "%$((POWERLEVEL9K_SHORTEN_DIR_LENGTH+1))(c:$POWERLEVEL9K_SHORTEN_DELIMITER/:)%${POWERLEVEL9K_SHORTEN_DIR_LENGTH}c")"
```

For this particular case, you can test it in a shortened version like so:

```zsh
print -P "%2(c:…/:)%1c"
```

And this will reliably fail the exact same way:

```zsh
> cd && print -P "%2(c:…/:)%1c"
~
> cd work/.. && print -P "%2(c:…/:)%1c"
…/~
```

I have a really hard time understanding how this statement works exactly, but I see if I can find out. [This document](http://zsh.sourceforge.net/Doc/Release/Prompt-Expansion.html) seems to be a good source.

**Comment by iFreilicht** [comment]
Ok this is definitely a bug in zsh. (I'm running version 5.0.2)

It comes down to the conditional expression not evaluating the way it should (I replaced the c with a ~ and moved the 2 inside here because it's clearer to me):

```zsh
> cd && print -P "%(2~:true:false)"
false
> cd work/.. && print -P "%(2~:true:false)"
true
```

This is a super-specific bug, because it only works if the argument 2. As soon as it's three (so as soon as your `POWERLEVEL9K_SHORTEN_DIR_LENGTH` is 2 or higher), it also returns `false` as it should.

Now to find out how to report bugs on zsh...

**Comment by Artistan** [comment]
Reporting ZSH issues here.
https://github.com/robbyrussell/oh-my-zsh/issues

**Comment by iFreilicht** [comment]
That's just the oh-my-zsh configuration framework. Bugs of zsh itself need to be reported via the mailing list as I found out. I already did that, and the bug has been confirmed, so it will hopefully be fixed in a future release.

I guess this issue here can be closed now.

**Comment by Artistan** [comment]
It would be great to tag the zsh issue here if you can find it on github.

**Comment by iFreilicht** [comment]
Zsh doesn't have an issue tracker on github, unfortunately, and their bug tracker on sourceforge is inactive.

They do have a mailing list archive, however. You can see the current state of the thread [here](http://www.zsh.org/mla/workers/2017/msg01808.html).

**Comment by Artistan** [comment]
@bhilburn - this can be closed since it is a ZSH issue, not a 9k issue.

**Comment by dritter** [comment]
This is already fixed via #773
